### ISSUE TITLE
Broken thru calibration

### ISSUE THREAD

**Issue by EB4FBZ** [ttrftech/NanoVNA#45]
Release 0.1.1 (commit d68ab20)

Bug found after updating my unit to this release. It worked fine with the stock firmware (GEN111.TAOBAO.COM). Double checked by flashing a backup.

After performing a full calibration (OPEN,SHORT,LOAD,ISO,THRU), the reference S21 shows a 0.5 to 1dB ripple. Ripple period depends on cable electrical length, but amplitude does not (tested with different quality cables).

 If only ISO+THRU or just THRU calibration is done, the trace is flat.

**Comment by EB4FBZ** [comment]
Same ripple present in 9508ea7. Attached photo.
![photo5787182004742238211](https://user-images.githubusercontent.com/55753450/65528671-1b6d1c00-def5-11e9-8e1a-ff4c273f0200.jpg)

**Comment by EB4FBZ** [comment]
More examples using both firmwares, allways using the original OSL kit but a quality RG316 30cm pigtail and quality adapters:

- First test: calibrating for reflection at the end of the cable, using a F-F SMA adapter. Removing the adapter for THRU calibration.

  - Original firmware: ![orig_S11calCable](https://user-images.githubusercontent.com/55753450/65624731-c3a0e480-dfca-11e9-86dc-42dadca8d3af.jpg)

  - Commit 9508ea7: ![0 1 1_S11calCable](https://user-images.githubusercontent.com/55753450/65624803-e0d5b300-dfca-11e9-85dd-16e6a59fd6e3.jpg)

- Second test: calibrating for reflection at the end of the cable, using a F-F SMA adapter. But this time i add a male-male adapter for THRU calibration.

  - Original firmware: ![orig_S11calCable_MM](https://user-images.githubusercontent.com/55753450/65624876-0498f900-dfcb-11e9-91b6-5a87e1231290.jpg)

  - Commit 9508ea7: ![0 1 1_S11calCable_MM](https://user-images.githubusercontent.com/55753450/65624898-0e226100-dfcb-11e9-8e73-448810e66344.jpg)

- Third test: calibrating for reflection right at the CH0 SMA. Adding the pigtail for THRU calibration.

  - Original firmware: ![orig_S11calCH0](https://user-images.githubusercontent.com/55753450/65624975-35792e00-dfcb-11e9-9281-cdb083b16ff5.jpg)

  - Commit 9508ea7: ![0 1 1_S11calCH0](https://user-images.githubusercontent.com/55753450/65624993-3e69ff80-dfcb-11e9-912f-6294499ec845.jpg)

It's clear to me that something weird is happening in newer firmwares. Some wrong interaction between reflection and transmission cal coeffs maybe? If i reset calibration and run a THRU cal without OSL steps, S21 response is perfectly flat.

**Comment by qrp73** [comment]
Cannot reproduce with the latest change set. 
Hardware: NanoVNA-H
CH0 to CH1 cable: 10 cm RG405, soldered SMA

I have 0.2 dB error at 0.05 MHz after calibration, but it is almost flat and smoothly going to 0 dB error at 900 MHz.

**Comment by qrp73** [comment]
When I use RESET and then CALIBRATE THRU only I get almost zero S21 error from 0.05 MHz to 900 MHz.

**Comment by qrp73** [comment]
I tried to split calibration into two stages:
1) OPEN, SHORT, LOAD, DONE
2) ISOLN, THRU, DONE

And got S21 error almost zero from 0.05 MHz to 900 MHz. So, it seems like temporary workaround for a proper calibration.

**Comment by EB4FBZ** [comment]
I think you are experiencing the same error, but with a shorter cable the ripple period is much longer and not so evident. 

You can't calibrate in two stages, starting a new calibration resets previous one.

**Comment by EB4FBZ** [comment]
More examples, this time including phase trace:

- Stock firmware with full calibration:
![orig_S11calCable_MM_phase](https://user-images.githubusercontent.com/55753450/65693834-34044000-e075-11e9-9abf-5a9c809d4d4c.jpg)

- Commit 9508ea7 with full calibration. OSL al the end of the cable and then add a male-male adapter for THRU:
![0 1 1_S11calCable_MM_phase](https://user-images.githubusercontent.com/55753450/65693850-38c8f400-e075-11e9-964e-138c967a32f3.jpg)

- Commit 9508ea7 with full calibration but reflection calibration just at the CH0 connector. Then add cable for THRU:
![0 1 1_S11calCH0_phase](https://user-images.githubusercontent.com/55753450/65693805-26e75100-e075-11e9-8592-dcfdd2252254.jpg)

- Commit 9508ea7 with THRU calibration only:
![0 1 1_THRUcal_phase](https://user-images.githubusercontent.com/55753450/65693796-22bb3380-e075-11e9-90af-42223dc99822.jpg)

Something is wrong.

**Comment by cho45** [comment]
This may caused by difference between NanoVNA and NanoVNA-H.

NanoVNA-H does not use ETERM_ES (source match) for calculating ETERM_ET (transmission).

NanoVNA: https://github.com/ttrftech/NanoVNA/blob/master/main.c#L1058
```
// CAUTION: Et is inversed for efficiency
static void
eterm_calc_et(void)
{
  int i;
  for (i = 0; i < sweep_points; i++) {
    // Et = 1/(S21mt - Ex)(1 - Es)
    float esr = 1 - cal_data[ETERM_ES][i][0];
    float esi = -cal_data[ETERM_ES][i][1];
    float s21mr = cal_data[CAL_THRU][i][0] - cal_data[CAL_ISOLN][i][0];
    float s21mi = cal_data[CAL_THRU][i][1] - cal_data[CAL_ISOLN][i][1];
    float etr = esr * s21mr - esi * s21mi;
    float eti = esr * s21mi + esi * s21mr;
    float sq = etr*etr + eti*eti;
    float invr = etr / sq;
    float invi = -eti / sq;
    cal_data[ETERM_ET][i][0] = invr;
    cal_data[ETERM_ET][i][1] = invi;
  }
  cal_status &= ~CALSTAT_THRU;
  cal_status |= CALSTAT_ET;
}
```

NanoVNA-H: https://github.com/hugen79/NanoVNA-H/blob/master/main.c#L1103 
```
// CAUTION: Et is inversed for efficiency
static void
eterm_calc_et(void)
{
  int i;
  for (i = 0; i < sweep_points; i++) {
    // Et = 1/(S21mt - Ex)(1 - Es)
   // float esr = 1 - cal_data[ETERM_ES][i][0];
    //float esi = -cal_data[ETERM_ES][i][1];
    float s21mr = cal_data[CAL_THRU][i][0] - cal_data[CAL_ISOLN][i][0];
    float s21mi = cal_data[CAL_THRU][i][1] - cal_data[CAL_ISOLN][i][1];
//    float etr = esr * s21mr - esi * s21mi;
//    float eti = esr * s21mi + esi * s21mr;
//    float sq = etr*etr + eti*eti;
//    float invr = etr / sq;
//    float invi = -eti / sq;
    float sq = s21mr*s21mr + s21mi*s21mi;
    float invr = s21mr / sq;
    float invi = -s21mi / sq;
    cal_data[ETERM_ET][i][0] = invr;
    cal_data[ETERM_ET][i][1] = invi;
  }
  cal_status &= ~CALSTAT_THRU;
  cal_status |= CALSTAT_ET;
}
```

**Comment by qrp73** [comment]
cho45, I also notice these changes in calibration calc, but I'm not sure - which version is better?

**Comment by edy555** [comment]
I found a failure. The expression in comment L1074 `// Et = 1/(S21mt - Ex)(1 - Es)` is incorrect, missing multiplier S11 for Es. The right is `Et = 1/(S21mt - Ex)(1 - Es S11)`.
And assume S11 is 1 in treating thru as ideal, it becomes `Et = 1/(S11mt - Ex)`.

In short, the fixed code in NanoVNA-H is right.

**Comment by edy555** [comment]
Just fixed at https://github.com/ttrftech/NanoVNA/commit/0b885f264078d11ab3f95ef74c9b6cd6aa146956

Please try recent master branch or binary in https://app.circleci.com/jobs/github/ttrftech/NanoVNA/84/artifacts

**Comment by EB4FBZ** [comment]
It's fixed! Thanks a lot @edy555

**Comment by qrp73** [comment]
yes, I can confirm, now error is smaller than 0.1 dB across entire bandwidth 0.05 - 900 MHz. For 0.05-300 even better - smaller than 0.01 dB :)

Thanks
